geomstats/visualization/grassmannian.py

```diff
@@ -0,0 +1,734 @@
+    return np.asarray(storage)
+
+
+def projection_to_three_d(projections, points):
+    """Apply projection matrix to a plotable 3D point.
+
+    Takes Grassmannian generated projection matrix and
+    converts it to a plotable 3D point.
+
+    INPUT: Array of 3x3 projection matricies or a single projection matrix
+
+    OUTPUT: Points on manifold but in a Euclidean representation
+    """
+    storage = []
+
+    vector = np.random.rand(3, 1)
+
+    if points == 1:
+        vector = np.matmul(projections, vector) / np.linalg.norm(
+            np.matmul(projections, vector)
+        )
+
+        return vector
+    for i in range(points):
+        x, y, z = np.matmul(projections[i], vector) / np.linalg.norm(
+            np.matmul(projections[i], vector)
+        )
+        vector = [x, y, z]
+        storage.append(vector)
+    return np.asarray(storage)
+
+
+def two_d_to_projection(vector):
+    """Take 2D point on manifold and gets its corresponding projection.
+
+    INPUT: Point on manifold as a 2x1 vector
+
+    OUTPUT: Projection matrix for point
+
+    WOULD BE USED FOR TANGENT REPRESENTATIONS
+    """
+    x = vector[0]
+    theta = np.arccos(x)
+    projection = np.zeros((2, 2))
+    projection[0][0] = (np.cos(theta)) ** 2
+    projection[0][1] = np.cos(theta) * np.sin(theta)
+    projection[1][0] = projection[0][1]
+    projection[1][1] = (np.sin(theta)) ** 2
+    return projection
+
+
+def submersion(point, k):
+    r"""Submersion that defines the Grassmann manifold.
+
+    The Grassmann manifold is defined here as embedded in the set of
+    symmetric matrices, as the pre-image of the function defined around the
+    projector on the space spanned by the first k columns of the identity
+    matrix by (see Exercise E.25 in [Pau07]_).
+    .. math:
+
+            \begin{pmatrix} I_k + A & B^T \\ B & D \end{pmatrix} \mapsto
+                (D - B(I_k + A)^{-1}B^T, A + A^2 + B^TB
+
+    This map is a submersion and its zero space is the set of orthogonal
+    rank-k projectors.
+
+    References
+    ----------
+    .. [Pau07]   Paulin, Frédéric. “Géométrie diﬀérentielle élémentaire,” 2007.
+                 https://www.imo.universite-paris-saclay.fr/~paulin
+                 /notescours/cours_geodiff.pdf.
+    """
+    _, eigvecs = gs.linalg.eigh(point)
+    eigvecs = gs.flip(eigvecs, -1)
+    flipped_point = Matrices.mul(Matrices.transpose(eigvecs), point, eigvecs)
+    b = flipped_point[..., k:, :k]
+    d = flipped_point[..., k:, k:]
+    a = flipped_point[..., :k, :k] - gs.eye(k)
+    first = d - Matrices.mul(
+        b, GeneralLinear.inverse(a + gs.eye(k)), Matrices.transpose(b)
+    )
+    second = a + Matrices.mul(a, a) + Matrices.mul(Matrices.transpose(b), b)
+    row_1 = gs.concatenate([first, gs.zeros_like(b)], axis=-1)
+    m_T = [Matrices.transpose(gs.zeros_like(b)), second]
+    row_2 = gs.concatenate(m_T, axis=-1)
+    return gs.concatenate([row_1, row_2], axis=-2)
+
+
+def _squared_d_g_a(point_a, point_b, metric):
```

Comment:
name this function `_squared_dist_grad_point_a` for consistency with next funciton